cloudify_cli/replace_certificates_config.py

```diff
@@ -0,0 +1,377 @@
+            for c in readq:
+                if c.recv_ready():
+                    output = stdout.channel.recv(len(c.in_buffer))
+                    self.logger.info(output)
+                    got_chunk = True
+                if c.recv_stderr_ready():
+                    output = stderr.channel.recv_stderr(
+                        len(c.in_stderr_buffer))
+                    got_chunk = True
+
+            if not got_chunk \
+                    and stdout.channel.exit_status_ready() \
+                    and not stderr.channel.recv_stderr_ready() \
+                    and not stdout.channel.recv_ready():
+                stdout.channel.shutdown_read()
+                stdout.channel.close()
+                break
+
+        stdout.close()
+        stderr.close()
+
+        exit_status = stdout.channel.recv_exit_status()
+        if exit_status != 0:
+            raise CloudifyCliError(output)
+
+        return output
+
+    @abc.abstractmethod
+    def _create_paths_mapping(self):
+        pass
+
+    def run_command(self, command):
+        try:
+            self.logger.debug('Running `{0}` on '
+                              '{1}'.format(command, self.host_ip))
+            return self._blocking_exec_command(command)
+        except CloudifyCliError as err:
+            self._command_failed(command, err.message)
+
+    @contextmanager
+    def sftp_connection(self):
+        sftp = None
+        try:
+            sftp = self.client.open_sftp()
+            yield sftp
+        finally:
+            if sftp:
+                sftp.close()
+
+    def put_file(self, local_path, remote_path):
+        self.logger.debug('Copying {0} to {1} on host {2}'.format(
+            local_path, remote_path, self.host_ip))
+        with self.sftp_connection() as sftp:
+            sftp.put(local_path, remote_path)
+
+    def _command_failed(self, command, err_msg):
+        self.errors_list.append('The command `{0}` on host {1} '
+                                'failed with error: '
+                                '{2}'.format(command, self.host_ip, err_msg))
+
+    def replace_certificates(self):
+        self.logger.info(('Replacing certificates on host '
+                          '{0}'.format(self.host_ip)))
+        self.run_command('cfy_manager replace-certificates -i '
+                         '{0}'.format(REMOTE_CONFIG_FILE_PATH))
+
+    def validate_certificates(self):
+        self.logger.info(('Validating certificates on host '
+                          '{0}'.format(self.host_ip)))
+        self._pass_certificates()
+        self.run_command('cfy_manager replace-certificates -i {0}'
+                         ' --only-validate'.format(REMOTE_CONFIG_FILE_PATH))
+
+    def needs_to_replace_certificates(self):
+        return any(self.certs_mapping.values())
+
+    def _pass_certificates(self):
+        self._prepare_new_certs_dir()
+        for cert_name, new_cert_path in self.certs_mapping.items():
```

Comment:
(not commenting about this line specifically but this line is the main path i guess)

I very much dislike the class-based stuff here, because all the classes take paths, and for the most part pass them through unchanged. I think the classes could be removed because they don't add much value.
The point is, to follow what happens to eg. `new_internal_cert_path`:
 - it is created in the config dict & file as `new_internal_cert_path`
 - it is loaded into `ReplaceCertificatesConfig`
 - `ReplaceCertificatesConfig` passes it - by name - into `ManagerNode`
 - `ManagerNode` receives it by name and sets it (well i gues the base class does) as `self.new_cert_path`
 - `_create_paths_mapping` runs and renames it back
 - the path is retrieved by name from `CERTS_MAPPING`
 - it is uploaded
 - it is put into the remote config file

see how many times i had to read `new_internal_cert_path`?! And there's like 10 of these files, which makes it hard to follow, because man my eyes start floating halfway deep into the stack. If somewhere you made a mistake and did `self.new_internal_cert = new_internal_key` then no way I'm gonna catch it :)
It's just reading the same name over and over but once in a while you throw a curveball and rename `new_internal_cert_path` to `new_cert_path` :D
(it doesn't help that all the names start with `new` and end with `path`, so they're more similar)

The main problem is that i must now hold several structures in my head: the config structure, the arguments to the Node classes, the dict output of the Node classes, the certs mapping, and the structure of the resulting remote config file.

I suggest reducing that count to just one item: the node config from the local config (the 
 item from `cluster_members`).
In there you have dicts with the following structure (example):
```python
{
    'new_internal_key_path': '',
    'new_external_cert_path': '',
    'new_external_key_path': '',
    'new_postgresql_client_cert_path': '',
    'new_postgresql_client_key_path': ''
}
```
So let's call a function `def upload_new_certs(host_ip, that_object):` which would: (pseudocode)
 - iterate over that dict, eg key=`new_internal_key_path`, value=`file.pem`
 - `conn = make_connection(host_ip)`
 - `local_path = config[key]` -
 - `remote_path = '/tmp/certs/{0}.pem'.format(key)` <- yes just that, no mapping
 - alternatively: `remote_path = '/tmp/certs/' + basename(local_path))` (same filename as local, doesnt really matter)
 - `conn.put_file(local_path, remote_path)`
 - put into the remote config file, `config[key] = remote_path` (no mapping)

it can still be written with classes if you want, or you can have several functions if you need specialcasing per node type, but the point is using the input config dict _directly_, without the N transformations to N object structures

Reply:
This is so true! Thanks, I also thought this way was too complicated but couldn't find a way to simplify it. Changed it now.